packages/react/src/components/ComposedModal/ComposedModal.tsx

```diff
@@ -81,194 +109,271 @@ ModalBody.propTypes = {
-    [`${prefix}--modal-container--${size}`]: size,
-    [`${prefix}--modal-container--full-width`]: isFullWidth,
-    [containerClassName]: containerClassName,
-  });
-
-  // Generate aria-label based on Modal Header label if one is not provided (L253)
-  let generatedAriaLabel;
-  const childrenWithProps = React.Children.toArray(children).map((child) => {
-    switch (child.type) {
-      case React.createElement(ModalHeader).type:
-        generatedAriaLabel = child.props.label;
-        return React.cloneElement(child, {
-          closeModal: closeModal,
-        });
-      case React.createElement(ModalFooter).type:
-        return React.cloneElement(child, {
-          closeModal: closeModal,
-          inputref: button,
+    function handleBlur({
+      target: oldActiveNode,
+      relatedTarget: currentActiveNode,
+    }) {
+      if (open && currentActiveNode && oldActiveNode && innerModal.current) {
+        const { current: bodyNode } = innerModal;
+        const { current: startSentinelNode } = startSentinel;
+        const { current: endSentinelNode } = endSentinel;
+        wrapFocus({
+          bodyNode,
+          startTrapNode: startSentinelNode,
+          endTrapNode: endSentinelNode,
+          currentActiveNode,
+          oldActiveNode,
+          selectorsFloatingMenus,
         });
-      default:
-        return child;
+      }
     }
-  });
 
-  useEffect(() => {
-    if (prevOpen !== isOpen) {
-      toggleClass(document.body, `${prefix}--body--with-modal-open`, isOpen);
+    function closeModal(evt) {
+      if (!onClose || onClose(evt) !== false) {
+        setisOpen(false);
+      }
     }
-  });
-
-  useEffect(() => {
-    return () =>
-      toggleClass(document.body, `${prefix}--body--with-modal-open`, false);
-  }, []); // eslint-disable-line react-hooks/exhaustive-deps
-
-  useEffect(() => {
-    toggleClass(document.body, `${prefix}--body--with-modal-open`, open);
-  }, [open, prefix]);
-
-  useEffect(() => {
-    const focusButton = (focusContainerElement) => {
-      if (focusContainerElement) {
-        const primaryFocusElement =
-          focusContainerElement.querySelector(selectorPrimaryFocus);
-        if (primaryFocusElement) {
-          primaryFocusElement.focus();
-          return;
+
+    const modalClass = cx(
+      `${prefix}--modal`,
+      isOpen && 'is-visible',
+      customClassName,
+      danger && `${prefix}--modal--danger`
+    );
+
+    const containerClass = cx(
+      `${prefix}--modal-container`,
+      size && `${prefix}--modal-container--${size}`,
+      isFullWidth && `${prefix}--modal-container--full-width`,
+      containerClassName
+    );
+
+    // Generate aria-label based on Modal Header label if one is not provided (L253)
+    let generatedAriaLabel;
+    const childrenWithProps = React.Children.toArray(children).map((child) => {
+      switch (true) {
+        case isElement(child) &&
+          child.type === React.createElement(ModalHeader).type: {
```

Comment:
This needed addition of `@types/react-is` dependency to give us the type-safety + we can now be sure about the actual type of the child (since `ReactNode` is a broad type